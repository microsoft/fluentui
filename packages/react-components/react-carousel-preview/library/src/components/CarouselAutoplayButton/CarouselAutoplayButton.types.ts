import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type CarouselAutoplayButtonSlots = {
  root: Slot<'div'>;
};

/**
 * CarouselAutoplayButton Props
 */
export type CarouselAutoplayButtonProps = ComponentProps<CarouselAutoplayButtonSlots> & {};

/**
 * State used in rendering CarouselAutoplayButton
 */
export type CarouselAutoplayButtonState = ComponentState<CarouselAutoplayButtonSlots>;
// TODO: Remove semicolon from previous line, uncomment next line, and provide union of props to pick from CarouselAutoplayButtonProps.
// & Required<Pick<CarouselAutoplayButtonProps, 'propName'>>
