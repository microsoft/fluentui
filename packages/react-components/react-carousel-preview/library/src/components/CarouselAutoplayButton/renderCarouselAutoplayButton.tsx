/** @jsxRuntime automatic */
/** @jsxImportSource @fluentui/react-jsx-runtime */

import { assertSlots } from '@fluentui/react-utilities';
import type { CarouselAutoplayButtonState, CarouselAutoplayButtonSlots } from './CarouselAutoplayButton.types';
import { renderToggleButton_unstable } from '@fluentui/react-button';

/**
 * Render the final JSX of CarouselAutoplayButton
 */
export const renderCarouselAutoplayButton_unstable = (state: CarouselAutoplayButtonState) => {
  assertSlots<CarouselAutoplayButtonSlots>(state);

  // We render the underlying react-button with injected carousel functionality
  return renderToggleButton_unstable(state);
};
