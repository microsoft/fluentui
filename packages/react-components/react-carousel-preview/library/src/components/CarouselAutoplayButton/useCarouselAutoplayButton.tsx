import * as React from 'react';
import { slot, useControllableState, useIsomorphicLayoutEffect } from '@fluentui/react-utilities';
import type { CarouselAutoplayButtonProps, CarouselAutoplayButtonState } from './CarouselAutoplayButton.types';
import { useToggleButton_unstable } from '@fluentui/react-button';
import { PlayCircleRegular, PauseCircleRegular } from '@fluentui/react-icons';
import { useEventCallback } from '@fluentui/react-utilities';
import { mergeCallbacks } from '@fluentui/react-utilities';
import { useCarouselContext_unstable as useCarouselContext } from '../CarouselContext';
import { ARIAButtonElement } from '@fluentui/react-aria';

/**
 * Create the state required to render CarouselAutoplayButton.
 *
 * The returned state can be modified with hooks such as useCarouselAutoplayButtonStyles_unstable,
 * before being passed to renderCarouselAutoplayButton_unstable.
 *
 * @param props - props from this instance of CarouselAutoplayButton
 * @param ref - reference to root HTMLDivElement of CarouselAutoplayButton
 */
export const useCarouselAutoplayButton_unstable = (
  props: CarouselAutoplayButtonProps,
  ref: React.Ref<ARIAButtonElement>,
): CarouselAutoplayButtonState => {
  const { onCheckedChange, checked, defaultChecked } = props;
  const [autoplay, setAutoplay] = useControllableState({
    state: checked,
    defaultState: defaultChecked,
    initialState: false,
  });

  const enableAutoplay = useCarouselContext(ctx => ctx.enableAutoplay);

  useIsomorphicLayoutEffect(() => {
    // Enable/disable autoplay on state change
    enableAutoplay(autoplay);
  }, [autoplay, enableAutoplay]);

  const handleClick = (event: React.MouseEvent<HTMLButtonElement & HTMLAnchorElement>) => {
    if (event.isDefaultPrevented()) {
      return;
    }
    const newValue = !autoplay;
    setAutoplay(newValue);
    onCheckedChange?.(event, { event, type: 'click', checked: newValue });
  };

  const handleButtonClick = useEventCallback(mergeCallbacks(handleClick, props.onClick));

  return {
    // We lean on react-button class to handle styling and icon enhancements
    ...useToggleButton_unstable(
      {
        icon: slot.optional(props.icon, {
          defaultProps: {
            children: autoplay ? <PauseCircleRegular /> : <PlayCircleRegular />,
          },
          renderByDefault: true,
          elementType: 'span',
        }),
        'aria-label': props.autoplayAriaLabel?.(autoplay),
        ...props,
        checked: autoplay,
        onClick: handleButtonClick,
      },
      ref as React.Ref<HTMLButtonElement>,
    ),
  };
};
