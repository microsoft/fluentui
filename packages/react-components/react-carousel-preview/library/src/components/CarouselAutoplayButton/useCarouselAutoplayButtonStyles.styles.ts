import { makeStyles, mergeClasses } from '@griffel/react';
import type { SlotClassNames } from '@fluentui/react-utilities';
import type { CarouselAutoplayButtonSlots, CarouselAutoplayButtonState } from './CarouselAutoplayButton.types';

export const carouselAutoplayButtonClassNames: SlotClassNames<CarouselAutoplayButtonSlots> = {
  root: 'fui-CarouselAutoplayButton',
  // TODO: add class names for all slots on CarouselAutoplayButtonSlots.
  // Should be of the form `<slotName>: 'fui-CarouselAutoplayButton__<slotName>`
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    // TODO Add default styles for the root element
  },

  // TODO add additional classes for different states and/or slots
});

/**
 * Apply styling to the CarouselAutoplayButton slots based on the state
 */
export const useCarouselAutoplayButtonStyles_unstable = (
  state: CarouselAutoplayButtonState,
): CarouselAutoplayButtonState => {
  const styles = useStyles();
  state.root.className = mergeClasses(carouselAutoplayButtonClassNames.root, styles.root, state.root.className);

  // TODO Add class names to slots, for example:
  // state.mySlot.className = mergeClasses(styles.mySlot, state.mySlot.className);

  return state;
};
