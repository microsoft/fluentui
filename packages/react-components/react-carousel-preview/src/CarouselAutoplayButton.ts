export * from './components/CarouselAutoplayButton/index';
