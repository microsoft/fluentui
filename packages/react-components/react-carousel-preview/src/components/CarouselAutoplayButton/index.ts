export * from './CarouselAutoplayButton';
export * from './CarouselAutoplayButton.types';
export * from './renderCarouselAutoplayButton';
export * from './useCarouselAutoplayButton';
export * from './useCarouselAutoplayButtonStyles.styles';
