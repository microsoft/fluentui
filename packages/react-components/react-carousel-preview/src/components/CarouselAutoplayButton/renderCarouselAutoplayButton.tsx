/** @jsxRuntime automatic */
/** @jsxImportSource @fluentui/react-jsx-runtime */

import { assertSlots } from '@fluentui/react-utilities';
import type { CarouselAutoplayButtonState, CarouselAutoplayButtonSlots } from './CarouselAutoplayButton.types';

/**
 * Render the final JSX of CarouselAutoplayButton
 */
export const renderCarouselAutoplayButton_unstable = (state: CarouselAutoplayButtonState) => {
  assertSlots<CarouselAutoplayButtonSlots>(state);

  // TODO Add additional slots in the appropriate place
  return <state.root />;
};
