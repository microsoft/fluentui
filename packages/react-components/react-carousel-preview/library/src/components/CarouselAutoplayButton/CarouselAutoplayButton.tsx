import * as React from 'react';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
import { useCarouselAutoplayButton_unstable } from './useCarouselAutoplayButton';
import { renderCarouselAutoplayButton_unstable } from './renderCarouselAutoplayButton';
import { useCarouselAutoplayButtonStyles_unstable } from './useCarouselAutoplayButtonStyles.styles';
import type { CarouselAutoplayButtonProps } from './CarouselAutoplayButton.types';

/**
 * If the Carousel is on auto-play, the user may opt into pausing the auto-play feature via the
 * CarouselAutoplayButton which must be present for auto-play to be enabled.
 *
 * If CarouselAutoplayButton is present, auto-play will default to true on mount.
 */
export const CarouselAutoplayButton: ForwardRefComponent<CarouselAutoplayButtonProps> = React.forwardRef(
  (props, ref) => {
    const state = useCarouselAutoplayButton_unstable(props, ref);

    useCarouselAutoplayButtonStyles_unstable(state);
    // TODO update types in packages/react-components/react-shared-contexts/src/CustomStyleHooksContext/CustomStyleHooksContext.ts
    // https://github.com/microsoft/fluentui/blob/master/rfcs/react-components/convergence/custom-styling.md
    // useCustomStyleHook_unstable('useCarouselAutoplayButtonStyles_unstable')(state);
    return renderCarouselAutoplayButton_unstable(state);
  },
);

CarouselAutoplayButton.displayName = 'CarouselAutoplayButton';
