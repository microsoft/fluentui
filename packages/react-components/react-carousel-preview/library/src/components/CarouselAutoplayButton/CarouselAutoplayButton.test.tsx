import * as React from 'react';
import { render } from '@testing-library/react';
import { isConformant } from '../../testing/isConformant';
import { CarouselAutoplayButton } from './CarouselAutoplayButton';
import { CarouselAutoplayButtonProps } from './CarouselAutoplayButton.types';

describe('CarouselAutoplayButton', () => {
  isConformant({
    Component: CarouselAutoplayButton as React.FunctionComponent<CarouselAutoplayButtonProps>,
    displayName: 'CarouselAutoplayButton',
    testOptions: {
      'has-static-classnames': [
        {
          props: {
            icon: 'Test Icon',
          },
        },
      ],
    },
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<CarouselAutoplayButton />);
    expect(result.container).toMatchSnapshot();
  });
});
