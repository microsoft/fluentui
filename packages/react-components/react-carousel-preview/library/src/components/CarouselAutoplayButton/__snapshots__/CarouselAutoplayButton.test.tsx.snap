// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselAutoplayButton renders a default state 1`] = `
<div>
  <div
    class="fui-CarouselAutoplayButton"
  >
    Default CarouselAutoplayButton
  </div>
</div>
`;
